Mathlib/Algebra/Category/AlgebraCat/Basic.lean

```diff
@@ -0,0 +1,251 @@
+
+initialize_simps_projections AlgebraCat (-isRing, -isAlgebra)
+
+namespace AlgebraCat
+
+instance : CoeSort (AlgebraCat R) (Type v) :=
+  ⟨AlgebraCat.carrier⟩
+
+attribute [coe] AlgebraCat.carrier
+
+instance : Category (AlgebraCat.{v} R) where
+  Hom A B := A →ₐ[R] B
+  id A := AlgHom.id R A
+  comp f g := g.comp f
+
+-- simplifies definitions below
+instance {S T : AlgebraCat.{v} R} : FunLike (S ⟶ T) S (fun _ => T) :=
+  ⟨fun f => f.toFun, fun _ _ h => AlgHom.ext (congr_fun h)⟩
+
+instance : ConcreteCategory.{v} (AlgebraCat.{v} R) where
+  forget :=
+    { obj := fun R => R
+      map := fun f => f.toFun }
+  forget_faithful := ⟨fun h => AlgHom.ext (by intros x; dsimp at h; rw [h])⟩
+
+instance {S : AlgebraCat.{v} R} : Ring ((forget (AlgebraCat R)).obj S) :=
+  (inferInstance : Ring S.carrier)
+
+instance {S : AlgebraCat.{v} R} : Algebra R ((forget (AlgebraCat R)).obj S) :=
+  (inferInstance : Algebra R S.carrier)
+
+instance hasForgetToRing : HasForget₂ (AlgebraCat.{v} R) RingCat.{v} where
+  forget₂ :=
+    { obj := fun A => RingCat.of A
+      map := fun f => RingCat.ofHom f.toRingHom }
+#align Algebra.has_forget_to_Ring AlgebraCat.hasForgetToRing
+
+instance hasForgetToModule : HasForget₂ (AlgebraCat.{v} R) (ModuleCat.{v} R) where
+  forget₂ :=
+    { obj := fun M => ModuleCat.of R M
+      map := fun f => ModuleCat.ofHom f.toLinearMap }
+#align Algebra.has_forget_to_Module AlgebraCat.hasForgetToModule
+
+/-- The object in the category of R-algebras associated to a type equipped with the appropriate
+typeclasses. -/
+def of (X : Type v) [Ring X] [Algebra R X] : AlgebraCat.{v} R :=
+  ⟨X⟩
+#align Algebra.of AlgebraCat.of
+
+/-- Typecheck a `alg_hom` as a morphism in `Algebra R`. -/
+def ofHom {R : Type u} [CommRing R] {X Y : Type v} [Ring X] [Algebra R X] [Ring Y] [Algebra R Y]
+    (f : X →ₐ[R] Y) : of R X ⟶ of R Y :=
+  f
+#align Algebra.of_hom AlgebraCat.ofHom
+
+@[simp]
+theorem ofHom_apply {R : Type u} [CommRing R] {X Y : Type v} [Ring X] [Algebra R X] [Ring Y]
+    [Algebra R Y] (f : X →ₐ[R] Y) (x : X) : ofHom f x = f x :=
+  rfl
+#align Algebra.of_hom_apply AlgebraCat.ofHom_apply
+
+instance : Inhabited (AlgebraCat R) :=
+  ⟨of R R⟩
+
+@[simp]
+theorem coe_of (X : Type u) [Ring X] [Algebra R X] : (of R X : Type u) = X :=
+  rfl
+#align Algebra.coe_of AlgebraCat.coe_of
+
+variable {R}
+
+/-- Forgetting to the underlying type and then building the bundled object returns the original
+algebra. -/
+@[simps]
+def ofSelfIso (M : AlgebraCat.{v} R) : AlgebraCat.of R M ≅ M where
+  hom := 𝟙 M
+  inv := 𝟙 M
+#align Algebra.of_self_iso AlgebraCat.ofSelfIso
+
+variable {M N U : ModuleCat.{v} R}
+
+@[simp]
+theorem id_apply (m : M) : (𝟙 M : M → M) m = m :=
+  rfl
+#align Algebra.id_apply AlgebraCat.id_apply
+
+@[simp]
+theorem coe_comp (f : M ⟶ N) (g : N ⟶ U) : (f ≫ g : M → U) = g ∘ f :=
+  rfl
+#align Algebra.coe_comp AlgebraCat.coe_comp
+
+variable (R)
+
+@[simp]
+theorem simps.free_obj_isRing_nsmul (a : ℕ) (b : FreeAlgebra R S) :
```

Comment:
should we get rid of these new simps lemmas since they aren't needed anymore?

Reply:
Yes please